Feat/storage based encryption key location 2 10.9

DeepDiver1975 opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.com/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next version of ownCloud.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" if the PR still has open tasks
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- test environment:
- test case 1:
- test case 2:
- ...

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

DeepDiver1975 committed:
feat: encryption keys location can be specified by the storage implementation

DeepDiver1975 committed:
feat: decrypt the stream of a file version as returned by IVersionedStorage

DeepDiver1975 committed:
fix: encryption storage wrapper will only delete keys if the storage has no own key handling

JammingBen committed:
Add unit tests for versioned storage methods

DeepDiver1975 committed:
fix: handle encryption keys with trashbin

DeepDiver1975 committed:
feat: add before_moveToTrash hook + add fileId to delete hook

DeepDiver1975 committed:
feat: adding new storage method getFileKey

DeepDiver1975 committed:
feat: adding new storage method setFileKey

DeepDiver1975 committed:
fix: trashbin pre_restore hook and encryption version on move

DeepDiver1975 committed:
feat: versioned storages are expected to handle retainVersions themselves

JammingBen committed:
Prevent encrypted files from being corrupted when overwriting them

JammingBen committed:
Add changelog item for #39623

JammingBen committed:
Fix getHeader unit tests

JammingBen committed:
Move changelog item #39623 to the released folder

micbar committed:
Bump version in version.php

micbar committed:
Update CHANGELOG

JammingBen committed:
Fix getting the file owner for share recipients

JammingBen committed:
Fix issue with wrong version authors being saved

JammingBen committed:
Merge pull request #39670 from owncloud/issues/39662

Fix getting the file owner for share recipients

JammingBen committed:
Merge pull request #39673 from owncloud/issues/39672

Fix issue with wrong version authors being saved

jnweiger committed:
bump version to 10.9.1 RC2

jnweiger committed:
Rename the last changelog folder, so that calens gets the date right, if running.

jnweiger committed:
bump version for final, update changelog via calens

jnweiger committed:
updated CHANGELOG via calens

DeepDiver1975 committed:
fix: handle file rename properly

DeepDiver1975 committed:
feat: copy trashed items to all owner trash bins

DeepDiver1975 committed:
feat: detect collisions of mounts wit local files

DeepDiver1975 committed:
feat: add absolute paths to rename hook

DeepDiver1975 committed:
fix: remove just uploaded files in case of error - also for storages which do not support part files

DeepDiver1975 committed:
fix: better error handling on file upload

DeepDiver1975 committed:
test: cleanup after failed upload

DeepDiver1975 commented:
it is honestly :hankey: that it is impossible to run ci on pr targeting older versions .... that's beyond the purpose of a ci system .... :angry:

phil-davis commented:
The CI "system" that existed for 10.9 no longer "exists". The drone secrets that were used then have been removed from the drone server and agents.
A similar thing happens if you try to run CI for an old oC10 from when we used Jenkins, or Travis, or... - in that case there is no Jenkins, Travis... server even online at all. So CI will have no chance to start.
The CI "scripts" that go with an old branch always depend on having a similar execution environment available for those "scripts".
If it is needed, then we can apply updated CI scripts and tools to an older branch and get it going.

DeepDiver1975 committed:
fix: also cleanup file cache on failed uploads

DeepDiver1975 committed:
fix: use empty() to check if header exists (port #40856)

IljaN committed:
feat: make restore abortable from within pre_restore (#41316)

Additionally this also passes the original restore location
of the inaccessible path before it is rewritten to the path
in home.
